### PR TITLE
Multiway match typechecking and translation for pattern guards

### PR DESCRIPTION
**Pull request by rajgodse**
Implements typechecking and translation for pattern guards.

The changes are mostly the straightforward generalization of the single case pattern guard construct.

The main complexity comes from changes related to the restructuring of `Typedtree.case`, whose field `c_rhs` used to be of type `expression`, but is now of type `case_rhs`, in conjunction with the way that multiway pattern guards change the notion of the rhs of a case.

The changes in ocamldoc are type-directed, but it is not apparent that they exhibit the intended behavior. It is worth trying to figure out the intended behavior of the modified functions through review.

### CODE REVIEW

File: ocaml/ocamldoc/odoc_ast.ml
Comment (ncik-roberts) [inline ocaml/ocamldoc/odoc_ast.ml]: As discussed, it's a bit weird to pass through `func_body` here as the only thing we do with `next_exp` is to check whether it's `Simple_rhs`. It might be clearer to pass through `None` here instead.

File: ocaml/ocamldoc/odoc_ast.ml
Comment (ncik-roberts) [inline ocaml/ocamldoc/odoc_ast.ml]: As discussed, it's probably clearer to keep the recursion over an expression, not an rhs.

File: ocaml/testsuite/tests/pattern-guards/test.ml
Comment (ncik-roberts) [inline ocaml/testsuite/tests/pattern-guards/test.ml]: Why is this test interesting? This is a genuine question, but also we should probably say in a comment.

File: ocaml/typing/parmatch.ml
Comment (ncik-roberts) [inline ocaml/typing/parmatch.ml:1856]: Nice instinct to enumerate the constructors of `c_rhs` :) (Feel free to resolve.)

File: ocaml/typing/parmatch.ml
Comment (ncik-roberts) [inline ocaml/typing/parmatch.ml]: Nit: should we rename to `is_refutation_case`? It's nice to give predicates names that start with copulas (or modal verbs).

File: ocaml/typing/typedtree.mli
Comment (ncik-roberts) [inline ocaml/typing/typedtree.mli:407]: Now that these fields are part of an inlined record argument to a constructor, we could consider scrubbing the `pg_*` and `bg_*` prefixes. (Type-based disambiguation will be principal, as the constructor name will always be enough to figure out where the field is coming from.) We don't need to do this, but I think it makes the names read a bit nicer, so it might be a good candidate for a small PR.

File: ocaml/typing/untypeast.ml
Comment (ncik-roberts) [inline ocaml/typing/untypeast.ml]:
```suggestion
    | Texp_function { arg_label; cases = [{c_lhs=p; c_rhs=Simple_rhs e}];
```

File: ocaml/lambda/translcore.ml
Comment (ncik-roberts) [inline ocaml/lambda/translcore.ml:284]: `exp_attributes` and `exp_extra` are doubly duplicated here. (They're being moved onto both the ident and the match expression, while also being preserved on the first case in the match's `cases`.) I really think this is a bug in the old code, and that these lists should be empty in the ident/match expression.

I'm not too concerned because `push_defaults` is getting deleted soon anyway. But we should leave a short comment explaining this weirdness, as it did temporarily take me aback.

File: ocaml/lambda/translcore.ml
Comment (ncik-roberts) [inline ocaml/lambda/translcore.ml]: `_impl` is an uninformative name. How about naming this one `event_function` and then naming the thing that's currently named `event_function` to `event_function_expr` to more strongly suggest a correspondence to `event_function_rhs`?

File: ocaml/lambda/translcore.ml
Comment (ncik-roberts) [inline ocaml/lambda/translcore.ml]: Oof, was this possibly exponential before? I didn't notice that `expr` was translated but then not being used in the pattern-guard case.

In any case, how would you feel about inlining `transl_body` at the places it's used? I think there are just two callsites, and that's not too offensive. Inlining can remove indirection, both semantically and literally.

File: ocaml/lambda/translcore.ml
Comment (ncik-roberts) [inline ocaml/lambda/translcore.ml]: The name should say `rhs` instead of `expr` now.

File: ocaml/lambda/translcore.ml
Comment (ncik-roberts) [inline ocaml/lambda/translcore.ml]: The name should say `rhs` instead of `expr` now.

File: ocaml/lambda/translcore.ml
Comment (ncik-roberts) [inline ocaml/lambda/translcore.ml:1646]: The `assert false` was never tripped before because:

```ocaml
# exception Exn;;
# let f x = match x with | Some x | exception x when x = Exn -> ();;
Error: Mixing value and exception patterns under when-guards is not supported.
```
Is this also true for when pattern guards?

  * If not, then the `assert false` could be tripped and we should fix that.
  * If so, we should write a test that evidences that.

File: ocaml/testsuite/tests/pattern-guards/test.ml
Comment (ncik-roberts) [inline ocaml/testsuite/tests/pattern-guards/test.ml]: This test file is getting very long. Should we consider splitting it up into multiple files? Using more comments to split it up?

```
(***********
 These tests show XYZ
*)
```

**Review by ncik-roberts — COMMENTED** [review]
The translcore changes are trickier than I thought, but they still seem fine. I left a few mostly-minor comments.

File: ocaml/ocamldoc/odoc_ast.ml
Comment (rajgodse) [reply ocaml/ocamldoc/odoc_ast.ml]: Done

File: ocaml/ocamldoc/odoc_ast.ml
Comment (rajgodse) [reply ocaml/ocamldoc/odoc_ast.ml]: Also done

File: ocaml/testsuite/tests/pattern-guards/test.ml
Comment (rajgodse) [reply ocaml/testsuite/tests/pattern-guards/test.ml]: It tests the change we made to `push_defaults`: in fact, it failed in the first version that I wrote (by failing to bind `s` at the right time but still using it).

Comment added!

File: ocaml/typing/typedtree.mli
Comment (rajgodse) [reply ocaml/typing/typedtree.mli:407]: Ah yeah that makes sense. I think a small PR after this one would be nice for this. I can also use this to simplify the Jane_syntax field names.

File: ocaml/lambda/translcore.ml
Comment (rajgodse) [reply ocaml/lambda/translcore.ml:284]: Got it, added a short comment explaining this

File: ocaml/lambda/translcore.ml
Comment (rajgodse) [reply ocaml/lambda/translcore.ml]: Yeah...kind of an unfortunate oversight, although perhaps it speaks to how having a nicer AST representation of the code is generally better for compiler development.

Inlining sounds good to me, it's a pretty short function body too.

File: ocaml/lambda/translcore.ml
Comment (rajgodse) [reply ocaml/lambda/translcore.ml:1646]: It is true, added such a test.

File: ocaml/testsuite/tests/pattern-guards/test.ml
Comment (rajgodse) [reply ocaml/testsuite/tests/pattern-guards/test.ml]: True, although we may be better-suited to do this on the parsing branch where we have already improved test readability. We could also try to merge the parsing branch and then modify this PR to merge into ocaml-flambda/flambda-backend/pattern-guards instead.

**Comment by rajgodse** [comment]
As discussed, rebased onto test changes in other PRs to consolidate tests before restructuring and splitting test cases into different test files.

**Comment by rajgodse** [comment]
Rebase onto new parsing

File: ocaml/ocamldoc/odoc_ast.ml
Comment (ncik-roberts) [inline ocaml/ocamldoc/odoc_ast.ml]: nit: could you undo the change in indentation to help with rebases?

File: ocaml/ocamldoc/odoc_ast.ml
Comment (ncik-roberts) [inline ocaml/ocamldoc/odoc_ast.ml]: nit: could you undo the change in indentation to help with rebases?

File: ocaml/ocamldoc/odoc_ast.ml
Comment (ncik-roberts) [inline ocaml/ocamldoc/odoc_ast.ml]: Nit: it helps to have short-cases earlier &mdash; that way, it's easier to see the scrutinee together with the patterns that match it on the same screenful of code. I'd consider moving this up.

File: ocaml/testsuite/tests/pattern-guards/nested_guards.ml
Comment (ncik-roberts) [inline ocaml/testsuite/tests/pattern-guards/nested_guards.ml]: I appreciate that you've split up these tests, but I suspect that this change would have made more sense in a separate PR so the squashed-and-merged git history is more intelligible. (And to make review a bit easier.) Would this be easy for you to do?

**Review by ncik-roberts — COMMENTED** [review]
A few fairly minor suggestions.

File: ocaml/ocamldoc/odoc_ast.ml
Comment (rajgodse) [reply ocaml/ocamldoc/odoc_ast.ml]: Makes sense, moved up to just before the last case.

File: ocaml/testsuite/tests/pattern-guards/nested_guards.ml
Comment (rajgodse) [reply ocaml/testsuite/tests/pattern-guards/nested_guards.ml]: This is pretty easy, partially because the changes are mostly logically separate from this PR's, which is also a good reason to make a separate PR for them. Done!

**Review by ncik-roberts — APPROVED** [review]
I think this will be good to merge after it's rebased and any conflicts are resolved. As discussed, we'll need to open a new PR from `multiway-match-parsing` into the flambda-backend repo even after this is merged.

**Pull request merged**
